src/Compilers/Core/MSBuildTask/LocateEditorConfigFiles.cs

```diff
@@ -0,0 +1,161 @@
+        }
+
+        public override bool Execute()
+        {
+            try
+            {
+                ExecuteCore();
+                return !Log.HasLoggedErrors;
+            }
+            catch (Exception ex)
+            {
+                // We'll explicitly log friendly errors and then rethrow in the case of IO errors while reading the file system.
+                // We'll only print the raw exception if we didn't already log something nice.
+                if (!Log.HasLoggedErrors)
+                {
+                    Log.LogErrorFromException(ex);
+                }
+
+                return false;
+            }
+        }
+
+        private void ExecuteCore()
+        {
+            var directoriesAddedToQueue = new HashSet<string>(StringComparer.Ordinal);
+            var directoriesToVisit = new Queue<DirectoryInfo>();
+            var editorConfigFiles = new List<ITaskItem>();
+
+            void addNewDirectoryIfNotAlreadyAdded(DirectoryInfo directory)
+            {
+                // If we're being passed files that don't exist, we'll just skip them and let the compiler handle the actual reporting.
+                if (directory.Exists)
+                {
+                    if (directoriesAddedToQueue.Add(directory.FullName))
+                    {
+                        directoriesToVisit.Enqueue(directory);
+                    }
+                }
+            }
+
+            foreach (var inputFile in InputFiles)
+            {
+                addNewDirectoryIfNotAlreadyAdded(new DirectoryInfo(Path.GetDirectoryName(Path.GetFullPath(inputFile.ItemSpec))));
+            }
+
+            while (directoriesToVisit.Count > 0)
+            {
+                var directory = directoriesToVisit.Dequeue();
+                var editorConfigFilePath = Path.Combine(directory.FullName, ".editorconfig");
+                bool isRootEditorConfig = false;
+
+                try
+                {
+                    if (File.Exists(editorConfigFilePath))
+                    {
+                        editorConfigFiles.Add(new TaskItem(editorConfigFilePath));
+
+                        // We will attempt to determine if this is a root .editorconfig, so we can stop enumerating farther if necessary.
+                        // We will do no further interpretation of the .editorconfig.
+
+                        isRootEditorConfig = FileIsRootEditorConfig(editorConfigFilePath);
+                    }
+                }
+                catch (IOException exception)
+                {
+                    // We'll log the error and continue; the "it failed" is handled in Execute().
+                    Log.LogErrorFromResources(nameof(ErrorString.General_UnableToReadFile), editorConfigFilePath, exception.Message);
+                }
+
+                if (!isRootEditorConfig && directory.Parent != null)
+                {
+                    addNewDirectoryIfNotAlreadyAdded(directory.Parent);
+                }
```

Comment:
Need to guard against `Parent` throwing here.  Consider the case where you're passed a directory path that you don't have access to.  Won't the `Parent` call throw `SecurityException`?  MSDN believes so.